---
title: "assignment"
author: "Maya Rassouli"
date: "2023-05-12"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(epiwraps)
  library(ggplot2)
  library(rGREAT)
})
```

### Download and decompress the data 

```{r, eval=FALSE}
options(timeout = 6000)
download.file("https://ethz-ins.org/content/w10.assignment.zip", "w10.assignment.zip")
unzip("w10.assignment.zip")
list.files()
```

### Use clustering and visualization to illustrate the relationship between the binding of the different proteins

## Prepare the regions and the tracks

```{r}
tracks <- list.files(pattern="bw$")
# focus on high-confidence peaks from each factor to define the universe of regions
peaks <- list.files(pattern="bed$")
# we first import the peaks
peaks <- lapply(peaks, rtracklayer::import.bed)
# we'll focus on the high-quality peaks
peaks <- lapply(peaks, FUN=function(x) x[x$score>800])
# we get the union of non-redundant regions
regions <- reduce(unlist(GRangesList(peaks)))
```

## Plot the signals 

```{r}
ml <- signal2Matrix(tracks, regions, extend=2000)

# we could do extra normalization if needed
# nf <- getNormFactors(tracks, useSeqLevels="chr1")
# ml <- rescaleSignalMatrices(ml, scaleFactors=nf)

plotEnrichedHeatmaps(ml, colors=c("white", "darkred"))

```

## Clustering

```{r}
# choose a meaningful number of clusters
cl2 <- clusterSignalMatrices(ml, k=2:10)
ggplot(cl2$varExplained, aes(k, varExplained)) + geom_line()

# cluster
set.seed(123) # ensure that it gives the same results everytime
cl <- clusterSignalMatrices(ml, k=6)
table(cl)
head(cl)
length(cl)
length(regions)
regions$cluster <- cl 
```

# Plot the clusters
```{r}
mycolors <- c("1"="red", "2"="blue", "3"="darkgreen", "4"="black", "5"="orange", "6"="purple") # add colors
plotEnrichedHeatmaps(ml, row_split=cl,  mean_color=mycolors)
```

# Plot the averages 
```{r}
d <- meltSignals(ml, splitBy=cl)
ggplot(d, aes(position, mean ,colour=sample)) + geom_line() + facet_wrap(~split)
```

### Use enrichment analysis (either GO or motif) on cluster 3

```{r}
# split the regions by cluster
split_regions <- split(regions, cl)
split_regions

# send the data to GREAT server to run an enrichment analysis
job <- submitGreatJob(gr=split_regions[["3"]], bg=regions, species="hg38")
res <- getEnrichmentTables(job)
names(res)
```

```{r}
# plot the top biological processes
bp <- res$`GO Biological Process`
head(bp)
ggplot(head(bp,15), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()
```

## Write a paragraph describing your results

From the clustering and its visualization: 
Cluster 1 is bound by CREB1, mainly, and CREB3L1. 
Cluster 2 is bound by CREB1, CREB3, CREB3L1. 
Cluster 3 is bound by CREB3. 
Cluster 4, 5 and 6 are bound by CREB1 and CREB3L1.

From the enrichment analysis: 
The enrichment map suggests that these three biological processes are found more often than expected by chances (higher hyperfold enrichment) : 
 - response to stimulus
 - cellular response to stimulus 
 - regulation of localization 
These regions are related to genes involved in response to stimulus (which is the most significant result).

